### PR TITLE
[Model] Add DGI Model

### PR DESCRIPTION
**Pull request by xavierzw**
## Description

This PR implement dgl version of DGI model according to the paper of "DEEP GRAPH INFOMAX" and its original code and test on three popular graph dataset cora, citeseer & pubmed.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes

Essentially, this PR uses DGL's API to implement DGI model.
And fix a small bug of "_normalize" function in data/citation_graph

### CODE REVIEW

**Comment by yzh119** [comment]
Thanks for your contribution! I'll take a look into this.

**Comment by yzh119** [comment]
Hi, I noticed a `best_dgi.pkl` in the folder. We recommend you to exclude all data/param files otherwise our repo would be too large.

**Comment by xavierzw** [comment]
I will remove this file. Thanks for reminding.

File: examples/pytorch/dgi/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+```
+
+```bash
+python train.py --dataset=pubmed --gpu=0
+```
+
+Results
+-------
+* cora: ~81.6 (81.2-82.1) (paper: 82.3)
+* citeseer: ~69.4 (paper: 71.8)
```
Comment (yzh119) [inline examples/pytorch/dgi/README.md]: This number seems to be lower than dgi paper. I'm trying to find the difference between your implementation and dgi official implementation.

File: python/dgl/data/citation_graph.py [hunk 2]
```diff
@@ -383,7 +383,7 @@ def _normalize(mx):
     """Row-normalize sparse matrix"""
     rowsum = np.array(mx.sum(1))
     r_inv = np.power(rowsum, -1).flatten()
-    r_inv[np.isinf(r_inv)] = np.inf
+    r_inv[np.isinf(r_inv)] = 0.
```
Comment (yzh119) [inline python/dgl/data/citation_graph.py]: Yes this is a bug. Thanks for fixing it.

File: examples/pytorch/dgi/README.md (hunk 1 above)
Comment (xavierzw) [reply examples/pytorch/dgi/README.md]: The citeseer's performance is much lower than dgi paper. I was trying to figure out, but not found the reason.

Comment (xavierzw) [reply examples/pytorch/dgi/README.md]: I did 5 recent experiments of citeseer. The result is between 69.3 and 71.8.
The detailed results are list below:
70.3
71.3
70.8
71.8
69.3

Comment (yzh119) [reply examples/pytorch/dgi/README.md]: I see, I've ran your code too and yield similar results. I also tried different initializers as Xavier in your Discriminator as the author did, but that did not influence the result.

**Comment by yzh119** [comment]
Everything looks good to me except for the performance on citeseer. We could merge this PR first.

**Comment by xavierzw** [comment]
> Everything looks good to me except for the performance on citeseer. We could merge this PR first.

Do I have the permission to merge the PR, or only the manager of the project have?

**Comment by yzh119** [comment]
Merged.
